store/tree.go

```diff
@@ -0,0 +1,480 @@
+type node struct {
+	tree  Tree
+	level uint16
+	key   []byte
+	// XXX: cache stored value?
+}
+
+type nodeIterator struct {
+	tree  Tree
+	level uint16
+	store.Iterator
+}
+
+func (iter nodeIterator) node() *node {
+	if !iter.Valid() {
+		return nil
+	}
+	res := node{
+		tree:  iter.tree,
+		level: iter.level,
+		key:   iter.Key()[7:],
+	}
+	return &res
+}
+
+func (node *node) children() (res children) {
+	bz := node.tree.store.Get(node.tree.nodeKey(node.level, node.key))
+	if bz != nil {
+		json.Unmarshal(bz, &res)
+	}
+	return
+}
+
+func (node *node) set(children children) {
+	bz, err := json.Marshal(children)
+	if err != nil {
+		panic(err)
+	}
+	node.tree.store.Set(node.tree.nodeKey(node.level, node.key), bz)
+}
+
+func (node *node) setLeaf(acc uint64) {
+	if node.level != 0 {
+		panic("setLeaf should not be called on branch node")
+	}
+	bz, err := json.Marshal(acc)
+	if err != nil {
+		panic(err)
+	}
+	node.tree.store.Set(node.tree.leafKey(node.key), bz)
+}
+
+func (node *node) delete() {
+	node.tree.store.Delete(node.tree.nodeKey(node.level, node.key))
+}
+
+func (node *node) leftSibling() *node {
+	return node.tree.nodeReverseIterator(node.level, nil, node.key).node()
+}
+
+func (node *node) rightSibling() *node {
+	iter := node.tree.nodeIterator(node.level, node.key, nil)
+	if !iter.Valid() {
+		return nil
+	}
+	if node.exists() {
+		// exclude node itself
+		iter.Next()
+	}
+	return iter.node()
+}
+
+func (node *node) child(n uint16) *node {
+	// TODO: set end to prefix iterator end
+	return node.tree.nodeIterator(node.level-1, node.children()[n].Index, nil).node()
+}
+
+func (node *node) parent() *node {
+	// first child inclusive case
+	parent := node.tree.nodeGet(node.level+1, node.key)
+	if parent.exists() {
+		return parent
+	}
+	parent = parent.leftSibling()
+	if parent.exists() {
+		return parent
+	}
+	return node.tree.nodeGet(node.level+1, nil)
+}
+
+func (node *node) exists() bool {
+	if node == nil {
+		return false
+	}
+	return node.tree.store.Has(node.tree.nodeKey(node.level, node.key))
+}
+
+func (node *node) updateAccumulation(c child) {
+	fmt.Printf("update on {%d %+v}\n", node.level, node.key)
+	if !node.exists() {
+		return // reached at the root
+	}
+
+	children := node.children()
+	idx, match := children.find(c.Index)
+	if !match {
+		panic("non existing key pushed from the child")
+	}
+	children = children.setAcc(idx, c.Acc)
+	node.set(children)
+	node.parent().updateAccumulation(child{node.key, children.accumulate()})
+}
+
+func (node *node) push(c child) {
+	fmt.Printf("push on {%d %+x}\n", node.level, node.key)
+	if !node.exists() {
+		node.create(children{c})
+		return
+	}
+
+	cs := node.children()
+	idx, match := cs.find(c.Index)
+	fmt.Printf("%d %v\n", idx, match)
+
+	// setting already existing child, move to updateAccumulation
+	if match {
+		node.updateAccumulation(c)
+		return
+	}
+
+	// inserting new child node
+	cs = cs.insert(idx, c)
+	parent := node.parent()
```

Comment:
I'm not clear on why this converges, because `parent` doesn't search for the closest key in the parents level, it instead just re-uses its own key. Shouldn't this cause every new element to make its own new parent nodes?

Reply:
parent() has three cases:
- if a node with the same key is at the upper level, use it as the parent
- if there is not, use the closest left key
- if there is not, it means this is the highest level and new parent node has to be made, which is with the nil key

the second case prevents the case you said, because if there a higher node whose start index is lesser than the current node's, it is detected and get returned.